services/flake_detection.py

```diff
@@ -0,0 +1,235 @@
+        )
+
+        potential_flake = self.curr_test_context.flake_dict[failure_message]
+        if potential_flake.counter > 1 and len(potential_flake.branch) > 2:
+            # Exact error happened on 2 other branches at least
+            self.curr_test_context.is_curr_flake = True
+            self.resulting_flakes[
+                self.curr_test_context.curr_test_id
+            ] = FlakeType.UnrelatedMatchingFailures
+
+            metrics.incr(
+                "flake_detection.detect_flakes.flake_detected",
+                1,
+                tags={"flake_type": str(FlakeType.UnrelatedMatchingFailures)},
+            )
+            return True
+        return False
+
+    @trace
+    def detect_flakes(self) -> List[Tuple[str, str]]:
+        """
+        Detect flaky tests on a given repo based on the test instances
+        gathered in the query in the constructor
+
+        A test is a flake if:
+        - it has failed on default branch
+        OR
+        - it has different outcomes on different runs for the same commit
+        OR
+        - it has the same failure messages on failures on unrelated branches
+
+        We check which kind of flake exists in that order
+
+
+        TODO: Differentiate between random flakes and infra flakes. Infra flakes
+        are flakes that are caused by a dependency or a temporary outage.
+        They can be defined as: matching failure messages on unrelated branches
+        clustered at a point in time, this test should only be failing for that
+        range in time.
+
+        Returns a list of tuples of flaky test id and type of flake
+        """
+        self.curr_test_context = FlakeDetectionContext()
+
+        metrics.distribution(
+            "flake_detection.detect_flakes.number_of_test_instances",
+            len(self.test_instances_ordered_by_test),
+            unit="test_instance",
+        )
+        with metrics.timing("flake_detection.detect_flakes.total_time_taken"):
+            for instance in self.test_instances_ordered_by_test:
+                with metrics.timing(
+                    "flake_detection.detect_flakes.process_individual_test_instance"
+                ):
+                    # because the query above orders by test_id, if we see a new test
+                    # we are now trying to determine if the next test is flaky
+                    if instance.test_id != self.curr_test_context.curr_test_id:
+                        self.curr_test_context.reset(instance.test_id)
+
+                    # if we've already determined the current test to be a flake
+                    # we don't have to keep examining instances of this test
+                    if self.curr_test_context.is_curr_flake:
+                        continue
+
+                    # check if failed on default branch
+                    # should probably automatically create an issue here
+                    if self.check_if_failed_on_default(instance):
```

Comment:
[nit] 

Move the metrics from inside the aux functions to inside each of the `if`s.
You'd still know what tag to emit, and the `for` won't look so... empty, the poor thing.
Also each function would be more focused on the conditions that mark a test as flake.

You can also change the `if...if...if` to `if...elif...elif` and keep the same effect of stopping earlier (if you have the metrics in the for) without all these `continue`s.